### PR TITLE
fix: read GOROOT at runtime, but fallback to compile-time

### LINKED ISSUE #46: depguard: create analyzer: couldn't expand $gostd: could not read GOROOT directory
1. build golangci-lint on computer A but execute it on computer B
2. computer B without $GOROOT env value，but has other go version

### PR DESCRIPTION
**Pull request by adamdecaf**
Fixes: https://github.com/OpenPeeDeeP/depguard/issues/46
More Info: https://github.com/golangci/golangci-lint/issues/3862

### CODE REVIEW

**Comment by ldez** [comment]
This PR doesn't fix the problem https://github.com/golangci/golangci-lint/issues/3862#issuecomment-1572893272

**Comment by adamdecaf** [comment]
I've updated this PR to include more strategies from https://github.com/golangci/golangci-lint/issues/3862#issuecomment-1572900443

File: internal/utils/variables.go [hunk 1]
```diff
@@ -45,7 +49,8 @@ func (e *gostdExpander) Expand() ([]string, error) {
 	if len(e.cache) != 0 {
 		return e.cache, nil
 	}
-	root := path.Join(build.Default.GOROOT, "src")
+	goroot := findGOROOT()
+	root := path.Join(goroot, "src")
```
Comment (ldez) [inline internal/utils/variables.go]:
```suggestion
	root := path.Join(findGOROOT(), "src")
```

File: internal/utils/variables.go [hunk 2]
```diff
@@ -61,6 +66,24 @@ func (e *gostdExpander) Expand() ([]string, error) {
 	return pkgPrefix, nil
 }
 
+func findGOROOT() string {
```
Comment (ldez) [inline internal/utils/variables.go:67]:
```suggestion
// code borrowed from https://github.com/golang/tools/blob/86c93e8732cce300d0270bce23117456ce92bb17/cmd/godoc/goroot.go#L15-L30
func findGOROOT() string {
```

**Review by ldez — CHANGES_REQUESTED** [review]

File: internal/utils/variables.go [hunk 3]
```diff
@@ -61,6 +66,24 @@ func (e *gostdExpander) Expand() ([]string, error) {
...
+		goroot = build.Default.GOROOT
+	}
+	if goroot == "" {
+		out, err := exec.Command("go", "env", "GOROOT").Output()
+		if err != nil {
+			return ""
+		}
+		goroot = strings.TrimSpace(string(out))
+	}
+	return filepath.Clean(goroot)
```
Comment (ldez) [inline internal/utils/variables.go]:
```suggestion
	if env := os.Getenv("GOROOT"); env != "" {
		return filepath.Clean(env)
	}
	def := filepath.Clean(runtime.GOROOT())
	if runtime.Compiler == "gccgo" {
		// gccgo has no real GOROOT, and it certainly doesn't
		// depend on the executable's location.
		return def
	}
	out, err := exec.Command("go", "env", "GOROOT").Output()
	if err != nil {
		return def
	}
	return strings.TrimSpace(string(out))
```

**Review by ldez — CHANGES_REQUESTED** [review]

**Review by gandarez — DISMISSED** [review]

**Comment by adamdecaf** [comment]
Updated from codereview, thanks.

**Review by dixonwille — APPROVED** [review]

**Comment by dixonwille** [comment]
Thanks you so much for this!

**Pull request merged**